pad_server/src/state.c

```diff
@@ -1,44 +1,67 @@
-#include "state.h"
+#include <errno.h>
 #include <pthread.h>
+#include <semaphore.h>
+#include <stdio.h>
+
+#include "state.h"
 
 /* TODO: docs */
 void padstate_init(padstate_t *state) {
-    pthread_mutex_init(&state->lock, NULL);
+    pthread_rwlockattr_setkind_np(
+        &state->rw_lock_attr, PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP); // Linux only, change if we deploy to
+                                                                             // nuttx, default for nuttx is writers pref
+    pthread_rwlock_init(&state->rw_lock, &state->rw_lock_attr);
+
     // TODO: Is this right? Can we assume if the program is running then the pad is armed?
     state->arm_level = ARMED_PAD;
     for (unsigned int i = 0; i < NUM_ACTUATORS; i++) {
         state->actuators[i] = false;
     }
 }
 
+//
+/* TODO: docs
+ *
+ */
+int padstate_get_level(padstate_t *state, arm_lvl_e *arm_val) {
+    int err;
+    err = pthread_rwlock_rdlock(&state->rw_lock);
+    if (err) return err;
+
+    *arm_val = state->arm_level;
+
+    err = pthread_rwlock_unlock(&state->rw_lock);
+    if (err) return err;
+
+    return 0;
```

Comment:
Could simplify this to `return pthread_rwlock_unlock(...)`. Or even just end with `return err;` and remove the if statement.